### PR TITLE
Feat/edge-model

### PR DESCRIPTION
**Pull request by MalachiteWind**
### Summary:
We add ensemble boostrap learning for edge points to compare linear and sinusoid regression fits

**Main Function:**
----
   - `regress_edge`: Main function that orchestrates the regression process, runs bootstrap trials, and generates plots for parameter histories and accuracy distributions.
   
 **Helper Functions:**
---
   - `_visualize_fits`: Generates a grid of subplots visualizing the regression fits for the top and bottom edges across multiple frames.
   - `create_sin_func` and `create_lin_func`: Helper functions to generate sinusoidal and linear regression functions.
   - `create_flat_data`: Flattens and prepares the data for regression.
   - `do_sinusoid_regression` and `do_lstsq_regression`: Perform regression using sinusoidal optimization or least-squares linear regression.
   - `bootstrap`: Conducts bootstrap sampling and regression for ensemble learning.
   - `ensem_regress_edge`: Manages the train/val split, calls the bootstrap function, and calculates accuracy metrics.
   - `plot_param_hist` and `plot_acc_hist`: Plot histograms of the learned regression parameters and accuracy results.
   - `_create_bs_idxs` and `_create_func_acc`: Utility functions for managing bootstrap indices and calculating accuracies.

*Edit*
I can't seem to figure out the python import ordering so it's throwing a lint error.

  
**MWE**
---

```
mitosis data regress_edge -F trials/debug2 -p data.filename=862pts -p regress_edge.train_len=default -p regress_edge.n_bags=default -p regress_edge.initial_guess=default  --debug
```

**Update** (sep 17, 2024)
----
- `initial_guess` parameter for optimization algorithm (method=Levenberg Marquardt) used to find sinusoid fit $y(r;t):=A \sin(\omega r - \gamma t) +B r$ inside `ensem_regress_edge` function now has optional default of `None`. 
- If `None`, $\omega$ and $\gamma$ are chosen randomly on unit circle, for initial guess. $B$ is selected to be the best linear fit parameter. $A$ is selected to bhe average error of points from linear growth rate.

### **Bimodal Issue**

The amplitude parameter of the sinusoid function, $A$, seems to show some bimodal behavior in the histogram of bootstrap bags:

image

Suggested fixes:
- Not so sure at the moment. Alex suggested adding in some additional parameters to the function being fit. *i.e.,*

$$y(r;t):= A \sin(\omega r - \gamma t +B) + Cr + D.$$
 
$$y(r;t):= A \sin(\omega r - \gamma t ) + B \cos(\omega r -\gamma t) + C r + D$$

### CODE REVIEW

File: plumex/tests/test_regression_pipeline.py [hunk 1]
```diff
@@ -123,3 +126,55 @@ def test_get_true_pred():
...
+        return a * np.sin(w * r - g * t) + b * r
+
+    axis = np.linspace(0, 1, 101)
+    tt, rr = np.meshgrid(axis, axis)
+
+    X = np.hstack((tt.reshape(-1, 1), rr.reshape(-1, 1)))
+    Y = sinusoid_func(tt, rr).reshape(-1)
+
+    step = 4
+    result = do_sinusoid_regression(X[::step], Y[::step], (1, 1, 1, 1))
```
Comment (Jacob-Stevens-Haas) [inline plumex/tests/test_regression_pipeline.py:144]: What's the point of `step` in these tests?

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Awesome, this is nice and readable!  For the linting, running `pre-commit run --all` and committing the result should work.  Is that failing?

Also, do you have an example of what the plots look like, or a sample invocation I can test out?

File: plumex/regress_edge.py [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+    plot_frameskip = len(center) / n_frames
+    frame_ids = [int(plot_frameskip * i) for i in range(n_frames)]
+
+    grid_size = int(np.ceil(np.sqrt(n_frames)))
+    
+    fig, axes = plt.subplots(grid_size, grid_size, figsize=(15, 15))
+    axes = axes.flatten()  
+
+    for i, idx in enumerate(frame_ids):
+        ax = axes[i]
```
[the remark this answers is not in the document]
Comment (MalachiteWind) [reply plumex/regress_edge.py]: not necessarily. if `n_frames` is not a perfect square they will differ.

File: plumex/tests/test_regression_pipeline.py (hunk 1 above)
Comment (MalachiteWind) [reply plumex/tests/test_regression_pipeline.py:145]: You know I'm not sure. Just testing regression capabilities on a subset of data.

Comment (MalachiteWind) [reply plumex/tests/test_regression_pipeline.py]: good point!! fixed.

**Pull request merged**